### ISSUE TITLE
TRACE is not thread-safe?

### ISSUE THREAD

**Issue by johnomotani** [boutproject/BOUT-dev#1511]
`TRACE` seems not to be thread-safe, and can cause segfaults if called from inside an OpenMP parallelized loop, see #1510. This is unfortunate because it means any function/method that might ever be called inside an OpenMP loop cannot use `TRACE`/`AUTO_TRACE`. I think this behaviour at least needs documenting. I don't know if there is a simple fix; could we perhaps make a separate `msg_stack` for each thread (maybe as a `std::vector` whose size is the maximum number of threads)?

**Comment by bendudson** [comment]
We could have a separate message stack for each thread, then the `BoutException` constructor could grab the message stack corresponding to the thread it is thrown in (or just store thread ID so this can be retrieved later). This may have some advantages:
 - No lock on the message stack
 - When a BoutException is thrown, the message stack corresponding to the thread which crashed could be printed, rather than a mix of all threads.

**Comment by dschwoerer** [comment]
I think it appeared in the get/setLocation branch:
https://travis-ci.org/boutproject/BOUT-dev/jobs/481860930
But I think it is not trivial to reproduce ...

**Comment by d7919** [comment]
Thanks!

**Comment by johnomotani** [comment]
@d7919 I could reproduce with f6bf4914613313061a024c7c17662b16f113cd07, `./configure --enable-openmp OMP_NUM_THREADS=2` in the tests `test-naulin-laplace` and `test-gyro`. I'm happy to try merging #1514 onto that commit and see if it fixes things (if it does I should revert the 'fix' in #1510 that just removes `AUTO_TRACE()`).

I found a compilation error with OpenMP enabled in #1514. I think I've fixed it (although I don't know the `pragma omp` statements well to be sure I did the right thing). Anyway, the updated #1514 does fix the segfaults in f6bf4914613313061a024c7c17662b16f113cd07. :+1:
